json_diff_test.go

```diff
@@ -427,3 +427,235 @@ func TestShouldReturnChangesWhenAObjectIsMissing(t *testing.T) {
+	}
+	if !reflect.DeepEqual(expectedNew, actualNew) {
+		t.Error("Test failed. Expected", expectedNew, "but returned", actualNew)
+	}
+}
+
+func TestShouldReturnOldAndNewValuesWhenHasDiffWithLessValue(t *testing.T) {
+	value := map[string]interface{}{
+		"data": map[string]interface{}{
+			"socio": map[string]interface{}{
+				"nome":           "Maria",
+				"documentoSocio": 1,
+				"tipo":           "ADM",
+			},
+		},
+		"str": "a",
+		"num": 1.0,
+	}
+	value2 := map[string]interface{}{
+		"data": map[string]interface{}{
+			"socio": map[string]interface{}{
+				"nome": "Maria",
+			},
+		},
+		"address": map[string]interface{}{
+			"street": map[string]interface{}{
+				"name": "Tom",
+			},
+		},
+		"str": "b",
+		"num": 1.0,
+	}
+
+	expectedDiff := []string{"address", "data.socio.documentoSocio", "data.socio.tipo", "str"}
+	expectedOld := map[string]interface{}{
+		"data": map[string]interface{}{
+			"socio": map[string]interface{}{"documentoSocio": 1, "tipo": "ADM"},
+		},
+		"str": "a",
+	}
+	expectedNew := map[string]interface{}{
+		"address": map[string]interface{}{
+			"street": map[string]interface{}{
+				"name": "Tom",
+			},
+		},
+		"str": "b",
+	}
+
+	actualDiff, actualOld, actualNew := jsondiff.DiffWithValues(value, value2)
+
+	if reflect.DeepEqual(actualOld, value) {
+		t.Error("Test failed. The actualOld and old can not be the same.", actualOld)
+	}
+	if reflect.DeepEqual(actualNew, value2) {
+		t.Error("Test failed. The actualNew and new can not be the same.", actualNew)
+	}
+	if !reflect.DeepEqual(expectedDiff, actualDiff) {
+		t.Error("Test failed. Expected", expectedDiff, "but returned", actualDiff)
+	}
+	if !reflect.DeepEqual(expectedOld, actualOld) {
+		t.Error("Test failed. Expected", expectedOld, "but returned", actualOld)
+	}
+	if !reflect.DeepEqual(expectedNew, actualNew) {
+		t.Error("Test failed. Expected", expectedNew, "but returned", actualNew)
+	}
+}
+
+func TestShouldReturnOldAndNewValuesWhenHasDiffInArrays(t *testing.T) {
+
+}
+
+func TestShouldReturnEmptyOldAndNewValuesWhenDoesNotHasDiff(t *testing.T) {
+	value := map[string]interface{}{
+		"num": 6.13,
+		"str": "a",
+	}
+
+	expectedDiff := []string{}
+	expectedOld := map[string]interface{}{}
+	expectedNew := map[string]interface{}{}
+
+	actualDiff, actualOld, actualNew := jsondiff.DiffWithValues(value, value)
+
+	if reflect.DeepEqual(actualOld, value) {
+		t.Error("Test failed. The actualOld and old can not be the same.", actualOld)
+	}
+	if reflect.DeepEqual(actualNew, value) {
+		t.Error("Test failed. The actualNew and new can not be the same.", actualNew)
+	}
+	if !reflect.DeepEqual(expectedDiff, actualDiff) {
+		t.Error("Test failed. Expected", expectedDiff, "but returned", actualDiff)
+	}
+	if !reflect.DeepEqual(expectedOld, actualOld) {
+		t.Error("Test failed. Expected", expectedOld, "but returned", actualOld)
+	}
+	if !reflect.DeepEqual(expectedNew, actualNew) {
+		t.Error("Test failed. Expected", expectedNew, "but returned", actualNew)
+	}
+}
```

Comment:
Could you write a test with json arrays too?

Reply:
done!